File: desktop/routes/mcx_resolver.js
/**
 * MCX Instrument Resolver for Desktop App
 * Downloads Upstox MCX instrument file and finds near-month futures.
 * Uses axios for reliable downloads with timeout and retry.
 */
const axios = require('axios');
const zlib = require('zlib');

const MCX_URL = 'https://assets.upstox.com/market-quote/instruments/exchange/MCX.json.gz';
const TARGET_COMMODITIES = ['CRUDEOIL', 'GOLD', 'SILVER'];

let _cache = {};
let _cacheTime = null;
const CACHE_HOURS = 4; // Refresh more often to catch contract rollovers

function findNearestFuture(instruments, underlying) {
  const now = Date.now();
  const futures = [];

  for (const inst of instruments) {
    const type = (inst.instrument_type || '').toUpperCase();
    if (type !== 'FUT' && type !== 'FUTCOM') continue;

    const sym = (inst.underlying_symbol || inst.asset_symbol || inst.name || '').toUpperCase();
    if (sym !== underlying.toUpperCase()) continue;

    const expRaw = inst.expiry;
    let expiry = null;
    if (typeof expRaw === 'number') {
      expiry = expRaw < 1e12 ? expRaw * 1000 : expRaw;
    } else if (typeof expRaw === 'string') {
      const parsed = parseInt(expRaw);
      if (!isNaN(parsed)) {
        expiry = parsed < 1e12 ? parsed * 1000 : parsed;
      } else {
        // Try date string parsing
        const d = new Date(expRaw);
        if (!isNaN(d.getTime())) expiry = d.getTime();
      }
    }

    if (expiry && expiry > now && inst.instrument_key) {
      futures.push({
        key: inst.instrument_key,
        symbol: inst.trading_symbol || '',
        expiry,
        lot: inst.lot_size || 1,
      });
    }
  }

  futures.sort((a, b) => a.expiry - b.expiry);
  return futures[0] || null;
}

async function downloadMCXInstruments() {
  try {
    console.log('[MCX] Downloading instrument file...');
    const resp = await axios.get(MCX_URL, {
      responseType: 'arraybuffer',
      timeout: 20000,
      headers: {
        'User-Agent': 'Mozilla/5.0',
        'Accept-Encoding': 'gzip',
      },
    });

    let data;
    try {
      // Try gunzip first (most likely)
      data = zlib.gunzipSync(Buffer.from(resp.data));
    } catch (_) {
      // Maybe it's already decompressed
      data = Buffer.from(resp.data);
    }

    const instruments = JSON.parse(data.toString('utf-8'));
    console.log(`[MCX] Downloaded ${instruments.length} instruments`);
    return instruments;
  } catch (e) {
    console.error('[MCX] Download failed:', e.message);
    return null;
  }
}

async function getMCXKeys() {
  // Check cache (only if non-empty)
  if (_cacheTime && _cache && Object.keys(_cache).length > 0) {
    const age = Date.now() - _cacheTime;
    if (age < CACHE_HOURS * 3600000) {
      return _cache;
    }
  }

  // Download and resolve
  const instruments = await downloadMCXInstruments();
  if (!instruments || instruments.length === 0) {
    console.warn('[MCX] No instruments downloaded, using existing cache');
    return _cache || {};
  }

  const result = {};
  for (const commodity of TARGET_COMMODITIES) {
    const fut = findNearestFuture(instruments, commodity);
    if (fut) {
      result[commodity.toLowerCase()] = fut.key;
      const expDate = new Date(fut.expiry).toLocaleDateString('en-IN');
      console.log(`[MCX] ${commodity}: ${fut.key} (${fut.symbol}, expiry: ${expDate})`);
    } else {
      console.warn(`[MCX] ${commodity}: No near-month future found`);
    }
  }

  if (Object.keys(result).length > 0) {
    _cache = result;
    _cacheTime = Date.now();
    console.log(`[MCX] Resolved ${Object.keys(result).length}/${TARGET_COMMODITIES.length} commodities`);
  } else {
    console.error('[MCX] Failed to resolve ANY commodity keys');
  }

  return result;
}

module.exports = { getMCXKeys };

File: backend/mcx_resolver.py
"""
MCX Instrument Resolver for Upstox API
Downloads Upstox instrument file and finds near-month futures for MCX commodities.
"""
import httpx
import csv
import io
import logging
from datetime import datetime, timezone, timedelta
from typing import Dict, Optional

logger = logging.getLogger(__name__)

MCX_INSTRUMENTS_URL = "https://assets.upstox.com/market-quote/instruments/exchange/MCX.csv.gz"
MCX_INSTRUMENTS_JSON_URL = "https://assets.upstox.com/market-quote/instruments/exchange/MCX.json.gz"

# Commodities we want to track
TARGET_COMMODITIES = ['CRUDEOIL', 'GOLD', 'SILVER']

# Cache
_mcx_cache: Dict[str, str] = {}
_cache_timestamp: Optional[datetime] = None
CACHE_DURATION = timedelta(hours=6)


async def _download_mcx_instruments() -> list:
    """Download MCX instrument list from Upstox"""
    instruments = []
    try:
        async with httpx.AsyncClient(timeout=30.0) as client:
            # Try JSON first
            resp = await client.get(MCX_INSTRUMENTS_JSON_URL)
            if resp.status_code == 200:
                import gzip
                import json
                data = gzip.decompress(resp.content)
                instruments = json.loads(data)
                logger.info(f"Downloaded {len(instruments)} MCX instruments (JSON)")
                return instruments
    except Exception as e:
        logger.warning(f"MCX JSON download failed: {e}")

    try:
        async with httpx.AsyncClient(timeout=30.0) as client:
            # Fallback to CSV
            resp = await client.get(MCX_INSTRUMENTS_URL)
            if resp.status_code == 200:
                import gzip
                data = gzip.decompress(resp.content)
                reader = csv.DictReader(io.StringIO(data.decode('utf-8')))
                instruments = list(reader)
                logger.info(f"Downloaded {len(instruments)} MCX instruments (CSV)")
                return instruments
    except Exception as e:
        logger.warning(f"MCX CSV download failed: {e}")

    return instruments


def _find_nearest_future(instruments: list, underlying: str) -> Optional[Dict]:
    """Find the near-month futures contract for a commodity"""
    now_ms = int(datetime.now(timezone.utc).timestamp() * 1000)
    futures = []

    for inst in instruments:
        inst_type = inst.get('instrument_type', '')
        symbol = inst.get('underlying_symbol', '') or inst.get('asset_symbol', '')

        if inst_type != 'FUT':
            continue
        if symbol.upper() != underlying.upper():
            continue

        # Parse expiry - could be milliseconds timestamp or date string
        expiry_raw = inst.get('expiry', '')
        expiry_ms = None
        if isinstance(expiry_raw, (int, float)):
            expiry_ms = int(expiry_raw)
        elif isinstance(expiry_raw, str):
            try:
                expiry_ms = int(expiry_raw)
            except ValueError:
                for fmt in ['%Y-%m-%d', '%d-%b-%Y']:
                    try:
                        expiry_ms = int(datetime.strptime(expiry_raw[:10], fmt).replace(tzinfo=timezone.utc).timestamp() * 1000)
                        break
                    except (ValueError, TypeError):
                        continue

        if expiry_ms and expiry_ms > now_ms:
            key = inst.get('instrument_key', '')
            trading_symbol = inst.get('trading_symbol', '')
            lot_size = inst.get('lot_size', 1)
            futures.append({
                'instrument_key': key,
                'trading_symbol': trading_symbol,
                'expiry_ms': expiry_ms,
                'lot_size': lot_size,
                'underlying': underlying,
            })

    if not futures:
        return None

    futures.sort(key=lambda x: x['expiry_ms'])
    return futures[0]


async def get_mcx_instrument_keys() -> Dict[str, str]:
    """
    Get instrument keys for near-month MCX commodity futures.
    Returns: {'crudeoil': 'MCX_FUT|...', 'gold': 'MCX_FUT|...', 'silver': 'MCX_FUT|...'}
    """
    global _mcx_cache, _cache_timestamp

    # Check cache
    if _cache_timestamp and (datetime.now(timezone.utc) - _cache_timestamp) < CACHE_DURATION and _mcx_cache:
        return _mcx_cache

    instruments = await _download_mcx_instruments()
    if not instruments:
        logger.warning("No MCX instruments downloaded, using cache or empty")
        return _mcx_cache or {}

    result = {}
    for commodity in TARGET_COMMODITIES:
        fut = _find_nearest_future(instruments, commodity)
        if fut and fut['instrument_key']:
            result[commodity.lower()] = fut['instrument_key']
            expiry_str = datetime.fromtimestamp(fut['expiry_ms'] / 1000, tz=timezone.utc).strftime('%d %b %Y')
            logger.info(f"MCX {commodity}: {fut['instrument_key']} ({fut['trading_symbol']}, expiry: {expiry_str})")
        else:
            logger.warning(f"MCX {commodity}: No near-month future found")

    if result:
        _mcx_cache = result
        _cache_timestamp = datetime.now(timezone.utc)

    return result


async def get_mcx_display_info() -> Dict:
    """Get MCX instrument info for display"""
    keys = await get_mcx_instrument_keys()
    return {
        'available': bool(keys),
        'instruments': keys,
        'count': len(keys),
    }
